btx/processing/pipca.py

```diff
@@ -0,0 +1,1011 @@
+            + ([rem_imgs % m] if rem_imgs % m else [])
+        )
+
+        # update model with remaining batches
+        for batch_size in batch_sizes:
+            self.fetch_and_update_model(batch_size)
+
+        if self.benchmarking:
+            self.save_interval_data()
+
+    def get_formatted_images(self, n, start_index, end_index):
+        """
+        Fetch n - x image segments from run, where x is the number of 'dead' images.
+
+        Parameters
+        ----------
+        n : int
+            number of images to retrieve
+        start_index : int
+            start index of subsection of data to retrieve
+        end_index : int
+            end index of subsection of data to retrieve
+
+        Returns
+        -------
+        ndarray, shape (end_index-start_index, n-x)
+            n-x retrieved image segments of dimension end_index-start_index
+        """
+
+        bin_factor = self.bin_factor
+        downsample = self.downsample
+
+        # may have to rewrite eventually when number of images becomes large,
+        # i.e. streamed setting, either that or downsample aggressively
+        imgs = self.psi.get_images(n, assemble=False)
+
+        if downsample:
+            imgs = bin_data(imgs, bin_factor)
+
+        imgs = imgs[
+            [i for i in range(imgs.shape[0]) if not np.isnan(imgs[i : i + 1]).any()]
+        ]
+
+        num_valid_imgs, p, x, y = imgs.shape
+        formatted_imgs = np.reshape(imgs, (num_valid_imgs, p * x * y)).T
+
+        return formatted_imgs[start_index:end_index, :]
+
+    def prime_model(self, X):
+        """
+        Initialiize model on sample of data using batch PCA.
+
+        Parameters
+        ----------
+        X : ndarray, shape (d x m)
+            set of m (d x 1) observations
+        """
+
+        if self.rank == 0:
+            print(f"Priming model with {self.num_components} samples...")
+
+        self.mu, self.total_variance = self.calculate_sample_mean_and_variance(X)
+
+        centered_data = X - np.tile(self.mu, self.num_components)
+        self.U, self.S, _ = np.linalg.svd(centered_data, full_matrices=False)
+
+        self.num_incorporated_images += self.num_components
+
+    def fetch_and_update_model(self, n):
+        """
+        Fetch images and update model.
+
+        Parameters
+        ----------
+        n : int
+            number of images to incorporate
+        """
+
+        rank = self.rank
+        start_index, end_index = self.split_indices[rank], self.split_indices[rank + 1]
+
+        img_batch = self.get_formatted_images(n, start_index, end_index)
+
+        self.update_model(img_batch)
+
+    def update_model(self, X):
+        """
+        Update model with new batch of observations using iPCA.
+
+        Parameters
+        ----------
+        X : ndarray, shape (d x m)
+            batch of m (d x 1) observations
+
+        Notes
+        -----
+        Implementation of iPCA algorithm from [1].
```

Comment:
Down the line, it'd be great to have an explanation of the underlying math in the Read the Docs.

Reply:
Opened #275

Reply:
Agreed!